Extend the `input` and `textarea` components to use the `dir` attribute

davidtrussler opened the pull request:
## What
Extends input and textarea components to accept a value for the `dir` attribute

## Why
In a previous piece of work (see[ Add RTL support for attachments](https://trello.com/c/RT5RbtmT/896-add-rtl-support-for-attachments)) we added support for right-to-left languages to the attachments pages on the Whitehall publishing app. This was done by adding a class to the parent of the `input` and `textarea` elements that would be used by CSS to render the text correctly. It became clear that it would be a more elegant and semantic solution to achieve this by use of the `dir` attribute of the elements. In addition this is recommended practice for instances where CSS is not loaded. 

Currently these components do not accept a value for that attribute so the changes here update these components accordingly. 

As part of this work it was also necessary to update the hint, label and error message components so that the correct value for the `dir` element can also be displayed on those elements. Another concern is to enable the element (`input` or `textarea`) and its corresponding `hint`, `label` and `error messages` to be able to render with different values for the `dir` attribute. This is for the specific use of the component in the Whitehall publishing app where a language is specified for a document. 

To this end the input and textarea components are considered with the following in mind: 
- the `dir` attribute is not set by default
- the `dir` attribute is set to the same value on both the `input`/`textarea` elements and their associated hint, label and error messages by default via the `right_to_left` boolean attribute of the component
- the `dir` attribute of the hint, label and error messages can optionally be set separately via the `right_to_left_field` boolean (for `input` or `textarea`) and `right_to_left_help` boolean (for hint, label and error messages) attributes of the component.

### Updated components

|element|example|
|---|---|
|hint|![Screenshot 2023-01-05 at 12 31 18](https://user-images.githubusercontent.com/6080548/210780908-5c06c0c0-eddf-4486-96ea-50d1fec43a68.png)|
|label|![Screenshot 2023-01-05 at 12 44 42](https://user-images.githubusercontent.com/6080548/210783849-b116a3ab-c9e4-4eac-93e0-c81645b11c26.png)|
|error message|![Screenshot 2023-01-05 at 12 50 42](https://user-images.githubusercontent.com/6080548/210784144-6b365dc4-e97a-46bc-8f29-b7dcaf3ee371.png)|
|error items|![Screenshot 2023-01-05 at 12 51 43](https://user-images.githubusercontent.com/6080548/210784312-01271be0-2b31-4438-b8ec-b765710c6967.png)|
|input|![Screenshot 2023-01-05 at 14 16 26](https://user-images.githubusercontent.com/6080548/210800817-4f6b73fb-0258-4efe-b4db-b58dfa6fb4e7.png)|
|input with separate `dir` attributes|![Screenshot 2023-01-05 at 14 22 16](https://user-images.githubusercontent.com/6080548/210802044-aaef06a5-b130-4d01-960a-1d216d1531c1.png)|
|textarea|![Screenshot 2023-01-05 at 14 18 19](https://user-images.githubusercontent.com/6080548/210801213-791720ab-8677-4c75-bd85-3b75597f4818.png)|
|textarea with separate `dir` attributes|![Screenshot 2023-01-05 at 14 19 09](https://user-images.githubusercontent.com/6080548/210801345-aa6c8252-95f6-4c7a-85b1-2783f2c350e7.png)|

andysellick commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@
 
 ## Unreleased
 
+* Extends input and textarea components to accept a value for the dir attribute ([PR #3081](https://github.com/alphagov/govuk_publishing_components/pull/3081))
```
Nitpick: can this match the PR name.

andysellick commented on app/views/govuk_publishing_components/components/_input.html.erb:
```diff
@@ -31,6 +31,8 @@
   prefix ||= nil
   suffix ||= nil
+  text_direction_rtl ||= false
+  text_direction_label_ltr ||= false
```
It might just be me, but reading this with one option referring to `rtl` and the other being `ltr` is a bit confusing. Is there a reason they're not both `_rtl` or both `_ltr`?

andysellick commented on app/views/govuk_publishing_components/components/docs/input.yml:
```diff
@@ -183,3 +183,17 @@ examples:
+
+      When `text_direction_rtl` is set to true both the input and the label are displayed right to left.
+
+      When `text_direction_label_ltr` is set to true the textarea is displayed right to left and the label is displayed left to right
```
Again this might just be me but I'm finding this a bit confusing. These two options both control the same things, but one overrides the other. I see that you want one flag to set everything to right to left and one flag to override that for the label - can you explain the use case for this? Would it be simpler to have one option for the label and one for the input?

There should also be a test that checks this behaviour as it doesn't look like it behaves as described. If I only set `text_direction_label_ltr` everything is displayed left to right - the textarea is not displayed right to left as described here.

andysellick commented on app/views/govuk_publishing_components/components/_textarea.html.erb:
```diff
@@ -17,6 +17,8 @@
   has_error ||= error_message || error_items.any?
   error_id = "error-#{SecureRandom.hex(4)}"
+  text_direction_rtl ||= false
+  text_direction_label_ltr ||= false
```
General suggestion - could the variable names used be clearer? Maybe use a general format like `text_direction_[element]_[direction]` e.g. `text_direction_input_right ||= false`. Later variables like `dir_textarea` could also use this convention for consistency. Distinguishing between `ltr` and `rtl` compared to `left` and `right` at a glance is harder.

andysellick reviewed (commented):
Looks like a great addition but the code is a little confusing and I think could do with some more tests in place.

davidtrussler commented:
@andysellick Thanks for the comments. In retrospect I think you may be right that the approach of defaulting both the `label` and the corresponding element to the same thing may cause more confusion than it solves. I'll think about that aspect further but my original thoughts were that for them to be the same would be the expected scenario and to set them differently would be an exception. But maybe that's not really the case.

kevindew reviewed (commented):
Nice job putting this together. 

I've got some suggestions on the syntax of calling this that I think could improve it (I hope I'm not contradicting any of your thoughts @andysellick, these are just loosely held opinions). I imagine there's some value in thrashing the syntax out a bit outside of implentation, since it's hard to change later.

I imagine we don't really need the full name `text_direction_rtl` for a boolean. I imagine we could go for `rtl` or `right_to_left` if we want a boolean, or dir if we accept a string (could be more flexible).

So I'd imagine if we did used `right_to_left` we could have:

```
<%= render "govuk_publishing_components/components/input", {
  name: "input",
  right_to_left: true
} %>
```

Result: We'd get a right to left input, no label.

If we did: 

```
<%= render "govuk_publishing_components/components/input", {
  label: {
    text: "My label"
  },
  name: "input",
  right_to_left: true
} %>
```

We'd get a right-to-left input with a right-to-left label.

And if we did:

```
<%= render "govuk_publishing_components/components/input", {
  label: {
    text: "My label",
    right_to_left: false
  },
  name: "input",
  right_to_left: true
} %>
```

We'd get a right-to-left input with a left-to-right label.

Equivalents with dir:

```
<%= render "govuk_publishing_components/components/input", {
  name: "input",
  dir: "rtl"
} %>
```

We'd get a right to left input, no label.
 
```
<%= render "govuk_publishing_components/components/input", {
  label: {
    text: "My label"
  },
  name: "input",
  dir: "rtl"
} %>
```

We'd get a right-to-left input with a right-to-left label.

```
<%= render "govuk_publishing_components/components/input", {
  label: {
    text: "My label",
    dir: "ltr"
  },
  name: "input",
  dir: "rtl"
} %>
```

We'd get a right-to-left input with a left-to-right label.

andysellick commented:
@kevindew niiiice. Yeah, go with that, that's much clearer.

I am still really interested to know what scenario would require a `rtl` input but a `ltr` label - can anyone enlighten me?

kevindew commented:
thanks, what do you think is better, a right to left Boolean or a text dir input?

And yes, I can understand that being surprising. It’s for publishing applications as the labels are expected to be in English.

So if you’re producing an Arabic translation of a document in Whitehall we want the user to input right to left, but the label of the field is right to left.

A thought did just occur to me about what to do about hints/error messages. As I guess they’ll remain left to right whatever is passed in hmm.

andysellick commented:
Ah that makes sense now, thanks.

I think the `right_to_left` boolean applied at the right point in the components options is clearest. That does depend on every component that's likely to need this to accept options in the same structured way (which might not be the case) but that's fixable.

I think this work is quite exciting - it could set a pattern for rolling out this functionality to any component that could need it in the future. Haven't thought much about the details, but I wonder if there's a way of making it easy to apply to any other components in the future? (might be a code thing, could be as simple as writing a short guide)

davidtrussler commented:
@andysellick I've updated this in line with the comments form yourself and @kevindew. 

I haven't done anything WRT the comment about other stuff like hints and error messages but the more I think about it the more I think that ought to be in here. It's not necessary for the purpose intended originally for these changes but it feels only half done without them now. Do you have any thoughts?

kevindew commented on app/views/govuk_publishing_components/components/_label.html.erb:
```diff
@@ -15,14 +16,16 @@
+  <%= tag.label text, id: id, for: html_for, class: css_classes, dir: dir %>
 <% end %>
 
 <% if hint_text.present? %>
```
I guess this hint text should also follow the right_to_left option.

davidtrussler replied:
Yes, good point, have added that. 👍

kevindew reviewed (commented):
@davidtrussler this looks great, thanks for digging into this further. I think you made the right choice adding it to form hint and error. 

What I'm now not totally sure on though is the best syntax to use to manage it. I think having label imply it is a bit surprising, since you wouldn't expect the label nested arguments to affect things not nested. I had a little play around and found it was probably easiest to allow fields to have two different right_to_left attributes, one for the field and one for the help. Both default to the standard right_to_left attribute if they're not set. 

This idea seemed hard to explain so I thought it was easiest communicated through a diff: https://github.com/alphagov/govuk_publishing_components/compare/7b5a7255b2c76b9592bce735a6dbeff40b89ae5a...rtl-kevin

Happy to have a chat over it

andysellick commented on app/views/govuk_publishing_components/components/_input.html.erb:
```diff
@@ -63,14 +64,19 @@
 
 <%= content_tag :div, class: form_group_css_classes do %>
   <% if label %>
+    <% label_rtl = label[:right_to_left] == false ? label[:right_to_left] : right_to_left %>
```
I think I understand this logic - if the label rtl is set to false, then use that, otherwise use the 'general' right_to_left value, is that right? It might be fractionally easier to read as shown below, but I notice that this logic repeats for the hint and error - we might have reached the tipping point where the amount of logic in the view justifies putting some of this code into a helper file, where code comments and a dedicated helper test would help clarify the functionality.

```
label_rtl = label[:right_to_left] == false ? false : right_to_left
```

andysellick replied:
Since this functionality is being applied to more than one component and could apply to others in future, it might be something to add to the shared helper.

davidtrussler replied:
This is now outdated and improved by the later changes along the lines suggested by Kevin.

andysellick commented on app/views/govuk_publishing_components/components/_input.html.erb:
```diff
@@ -84,17 +90,23 @@
   <% end %>
 
   <% if has_error %>
+    <% error_rtl = label_rtl.nil? ? right_to_left : label_rtl %>
```
Asking mainly for my own understanding (it's early, my brain's still waking up) - the logic for hint and error differs from the logic from label? What's the difference there?

davidtrussler replied:
This is now outdated and improved by the later changes along the lines suggested by Kevin.

andysellick commented on app/views/govuk_publishing_components/components/docs/error_message.yml:
```diff
@@ -15,9 +15,22 @@ examples:
       text: "Please enter your National Insurance Number"
       id: "error_id"
+  with_dir_attribute:
+    description: Error items are a common pattern where a collection of error is passed with each item having a text attribute of the error
```
This description and the one on the `with_items_and_dir_attribute` example are the same, and don't describe what is being demonstrated.

davidtrussler replied:
Oops, that was a bit of lazy copy and paste. Fixed now 👍

andysellick reviewed (commented):
A few more thoughts and questions. Also can some of these commits be combined? Ideally having the overall change to each component in a single commit would be good (i.e. change to template, test and documentation).

davidtrussler commented:
> @davidtrussler this looks great, thanks for digging into this further. I think you made the right choice adding it to form hint and error.
> 
> What I'm now not totally sure on though is the best syntax to use to manage it. I think having label imply it is a bit surprising, since you wouldn't expect the label nested arguments to affect things not nested. I had a little play around and found it was probably easiest to allow fields to have two different right_to_left attributes, one for the field and one for the help. Both default to the standard right_to_left attribute if they're not set.
> 
> This idea seemed hard to explain so I thought it was easiest communicated through a diff: [7b5a725...rtl-kevin](https://github.com/alphagov/govuk_publishing_components/compare/7b5a7255b2c76b9592bce735a6dbeff40b89ae5a...rtl-kevin)
> 
> Happy to have a chat over it

@kevindew I've updated the input component in line with what you have suggested (in [this commit](https://github.com/alphagov/govuk_publishing_components/pull/3081/commits/01ddd227120041da1ec007f5ff9e95059548d735)), though I've changed that a bit to hopefully be pretty clear to the user. This gives two new variations on the component documentation as below. If that looks OK I'll extend it to the textarea component. 

|Single right-to-left attribute|Multiple right-to-left attributes|
|-|-|
|![Screenshot 2022-12-22 at 12 58 20](https://user-images.githubusercontent.com/6080548/209139365-63d654ec-f8e4-4c8d-9843-7dd1a7de4b92.png)|![Screenshot 2022-12-22 at 12 59 02](https://user-images.githubusercontent.com/6080548/209139464-901ee973-1e14-4b32-ac98-b1c518f85eaa.png)|

kevindew commented:
Thanks @davidtrussler - that looks good to me 👍

andysellick commented on app/assets/stylesheets/govuk_publishing_components/components/_hint.scss:
```diff
@@ -1 +1,5 @@
 @import "govuk/components/hint/hint";
+
+.govuk-hint {
+  display: block;
```
Is this needed? `.govuk-hint` is a div element, which is block by default.

davidtrussler replied:
That's interesting, actually when the `.govuk-hint` class is used in the label component (though not in the other classes) it is a span. If it is an inline element the text floats to the left and looks wrong (which was the reason for adding it). I've changed it to only apply where it is used in the label component.

davidtrussler replied:
Yes, I agree, that would be the better solution. I've updated that to a div and removed the addition to the CSS.

andysellick commented on app/views/govuk_publishing_components/components/docs/label.yml:
```diff
@@ -65,3 +65,12 @@ examples:
       html_for: "id-that-matches-input-6"
+  with_dir_attribute:
+    description: |
+      To allow the correct display of right to left languages.
```
Maybe add something like `Dir also applies to label hint, as shown.`?

davidtrussler replied:
Yes, good point. Have added that. 👍

andysellick commented on app/views/govuk_publishing_components/components/docs/error_message.yml:
```diff
@@ -15,9 +15,22 @@ examples:
     data:
       text: "Please enter your National Insurance Number"
       id: "error_id"
+  with_dir_attribute:
```
Nitpick: I'd put this example beneath the `with_items` example, as `with_items` feels like more of a prominent use case.

davidtrussler replied:
Sounds like a fair point. Have changed that. 👍

andysellick commented on app/views/govuk_publishing_components/components/_input.html.erb:
```diff
@@ -107,6 +113,7 @@
     autofocus: autofocus,
     class: css_classes,
     data: data,
+    dir: right_to_left_field ? "rtl" : nil,
```
I might not be fully understanding this, so please feel free to correct me - there are three rtl options: the normal one, `field` and `help`. `help` is applied to the label, hint and error message, and `field` is applied to the input itself. But the normal one isn't applied directly to any elements, it's just used to determine the others, is that right?

If that's correct, could `right_to_left` not be directly applied to the input, and `field` removed? 

```Ruby
right_to_left ||= false
right_to_left_help = right_to_left_help.nil? ? right_to_left : right_to_left_help

# label: right_to_left_help
# hint: right_to_left_help
# error: right_to_left_help
# input: right_to_left ? "rtl" : nil
```

davidtrussler replied:
Hmm, yes, that makes sense. And will apply equally to the textarea component. I've updated the input and textarea components to reflect this with changes to the example documentation, the html and the test.

andysellick reviewed (commented):
Looking much better, good job. Just a few minor things and one thing that is probably me not understanding correctly.

davidtrussler commented:
> Looking much better, good job. Just a few minor things and one thing that is probably me not understanding correctly.

Thanks @andysellick - I think I've covered the points you've raised if you could give it a look over 😄

andysellick reviewed (approved):
Nice work. I also like the clean commit structure.

I was wondering if this is needed again in the future whether it would be possible to implement some kind of shared code for adding this to components, but I think that's out of scope for now.
